kvdb/etcd/stm.go

```diff
@@ -320,42 +345,224 @@ func runSTM(s *stm, apply func(STM) error) error {
 	return executeErr
 }
 
-// add inserts a txn response to the read set. This is useful when the txn
-// fails due to conflict where the txn response can be used to prefetch
-// key/values.
-func (rs readSet) add(txnResp *v3.TxnResponse) {
-	for _, resp := range txnResp.Responses {
+func newReadSet() *readSet {
+	return &readSet{
+		tree:       btree.New(2),
+		fullRanges: make(map[string]bool),
+	}
+}
+
+// add inserts key/values to to read set.
+func (rs *readSet) add(responses []*pb.ResponseOp) {
+	for _, resp := range responses {
 		getResp := (*v3.GetResponse)(resp.GetResponseRange())
 		for _, kv := range getResp.Kvs {
-			rs[string(kv.Key)] = stmGet{
-				val: string(kv.Value),
-				rev: kv.ModRevision,
+			rs.addItem(
+				string(kv.Key), string(kv.Value), kv.ModRevision,
+			)
+		}
+	}
+}
+
+// addFullRange adds all full ranges to the read set.
+func (rs *readSet) addFullRange(prefixes []string, responses []*pb.ResponseOp) {
+	for i, resp := range responses {
+		getResp := (*v3.GetResponse)(resp.GetResponseRange())
+		for _, kv := range getResp.Kvs {
+			rs.addItem(
+				string(kv.Key), string(kv.Value), kv.ModRevision,
+			)
+		}
+
+		rs.fullRanges[prefixes[i]] = true
+	}
+}
+
+// presetItem presets a key to zero revision if not already present in the read
+// set.
+func (rs *readSet) presetItem(key string) {
+	item := &stmGet{
+		KV: KV{
+			key: key,
+		},
+		rev: 0,
+	}
+
+	if !rs.tree.Has(item) {
+		rs.tree.ReplaceOrInsert(item)
+	}
+}
+
+// addItem adds a single new key/value to the read set (if not already present).
+func (rs *readSet) addItem(key, val string, modRevision int64) {
+	item := &stmGet{
+		KV: KV{
+			key: key,
+			val: val,
+		},
+		rev: modRevision,
+	}
+
+	rs.tree.ReplaceOrInsert(item)
+}
+
+// hasFullRange checks if the read set has a full range prefetched.
+func (rs *readSet) hasFullRange(prefix string) bool {
+	return rs.fullRanges[prefix]
+}
+
+// next returns the pre-fetched next value of the prefix. If matchKey is true,
+// it'll simply return the key/value that matches the passed key.
+func (rs *readSet) next(prefix, key string, matchKey bool) (*stmGet, bool) {
+	pivot := &stmGet{
+		KV: KV{
+			key: key,
+		},
+	}
+
+	var result *stmGet
+	rs.tree.AscendGreaterOrEqual(
+		pivot,
+		func(item btree.Item) bool {
+			next := item.(*stmGet)
+			if !matchKey && next.key == key {
+				return true
+			}
+
+			if next.rev != 0 && strings.HasPrefix(next.key, prefix) {
```

Comment:
In this context, `rev` 0 means that this key is _freshly_ written in this transaction context right? So me writing a new key in the invoice registry as an example.

Reply:
So revision 0 means that we read the key (or the whole range with the same prefix) but there wasn't anything in the DB with that key. By "taking note" we later on (when attempting to commit) can add the key the the compare set to ensure that the key still isn't present in the DB.